tests/templates/test_subroutines/test_prepselprep.py

```diff
@@ -0,0 +1,265 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# pylint: disable=protected-access
+"""
+Tests for the PrepSelPrep template.
+"""
+import copy
+
+import numpy as np
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as pnp
+
+
+def test_standard_checks():
+    """Run standard validity tests."""
+    lcu = qml.ops.LinearCombination([0.25, 0.75], [qml.Z(2), qml.X(1) @ qml.X(2)])
+    control = [0]
+
+    op = qml.PrepSelPrep(lcu, control)
+    qml.ops.functions.assert_valid(op)
+
+
+def test_repr():
+    """Test the repr method."""
+    lcu = qml.dot([0.25, 0.75], [qml.Z(2), qml.X(1) @ qml.X(2)])
+    control = [0]
+
+    op = qml.PrepSelPrep(lcu, control)
+    assert (
+        repr(op)
+        == "PrepSelPrep(coeffs=(0.25, 0.75), ops=(Z(2), X(1) @ X(2)), control=<Wires = [0]>)"
+    )
+
+
+class TestPrepSelPrep:
+    """Test the correctness of the decomposition"""
+
+    def manual_circuit(self, lcu, control):
+        """Circuit equivalent to decomposition of PrepSelPrep"""
+
+        coeffs, unitaries = lcu.terms()
+        normalized_coeffs = qml.math.sqrt(qml.math.abs(coeffs)) / qml.math.norm(
+            qml.math.sqrt(qml.math.abs(coeffs))
+        )
+
+        qml.StatePrep(normalized_coeffs, wires=control)
+        qml.Select(unitaries, control=control)
+        qml.adjoint(qml.StatePrep(normalized_coeffs, wires=control))
+
+        return qml.state()
+
+    def prepselprep_circuit(self, lcu, control):
+        """PrepSelPrep circuit used for testing"""
+
+        qml.PrepSelPrep(lcu, control)
+        return qml.state()
+
+    def test_decomposition(self):
+        """Test that the template produces the corrent decomposition"""
+
+        dev = qml.device("default.qubit")
+        manual = qml.QNode(self.manual_circuit, dev)
+        prepselprep = qml.QNode(self.prepselprep_circuit, dev)
+
+        lcu = qml.dot([0.25, 0.75], [qml.Z(2), qml.X(1) @ qml.X(2)])
+        assert np.array_equal(
+            qml.matrix(manual, wire_order=[0, 1, 2])(lcu, control=0),
+            qml.matrix(prepselprep, wire_order=[0, 1, 2])(lcu, control=0),
+        )
+
+        lcu = qml.dot([1 / 2, 1 / 2], [qml.Identity(0), qml.PauliZ(0)])
+        assert np.array_equal(
+            qml.matrix(manual, wire_order=[0, "ancilla"])(lcu, control="ancilla"),
+            qml.matrix(prepselprep, wire_order=[0, "ancilla"])(lcu, control="ancilla"),
+        )
+
+        a = 0.25
+        b = 0.75
+        A = np.array([[a, 0, 0, b], [0, -a, b, 0], [0, b, a, 0], [b, 0, 0, -a]])
+        lcu = qml.pauli_decompose(A)
+        coeffs, unitaries = lcu.terms()
+        unitaries = [qml.map_wires(op, {0: 1, 1: 2}) for op in unitaries]
+        lcu = qml.dot(coeffs, unitaries)
+        assert np.array_equal(
+            qml.matrix(manual, wire_order=[0, 1, 2])(lcu, control=0),
+            qml.matrix(prepselprep, wire_order=[0, 1, 2])(lcu, control=0),
+        )
```

Comment:
We should definitely be testing multiple cases of different LCUs as you are here, but the python best practice for writing tests like these is to make use of the `@pytest.mark.parametrize` decorator. Could you reformat any of your tests that have this multiple case behaviour to make use of that best practice.